pulser/simulation/simresults.py

```diff
@@ -14,63 +14,319 @@
 """Classes for containing and processing the results of a simulation."""
 
 from __future__ import annotations
-
-from typing import Optional, Union
+from collections import Counter
+from abc import ABC, abstractmethod
+from typing import Optional, Union, cast, Tuple
 from collections.abc import Sequence
 
+import matplotlib.pyplot as plt
 import qutip
+from qutip.piqs import isdiagonal
 import numpy as np
 from numpy.typing import ArrayLike
 
 
-class SimulationResults:
+class SimulationResults(ABC):
     """Results of a simulation run of a pulse sequence.
 
+    Parent class for NoisyResults and CleanResults.
     Contains methods for studying the states and extracting useful information
     from them.
     """
 
-    def __init__(self, run_output: Sequence[qutip.Qobj], dim: int, size: int,
-                 basis_name: str, meas_basis: Optional[str] = None) -> None:
+    def __init__(self, dim: int, size: int, basis_name: str,
+                 sim_times: ArrayLike) -> None:
         """Initializes a new SimulationResults instance.
 
         Args:
-            run_output (list[qutip.Qobj]): List of ``qutip.Qobj`` corresponding
-                to the states at each time step after the evolution has been
-                simulated.
             dim (int): The dimension of the local space of each atom (2 or 3).
             size (int): The number of atoms in the register.
             basis_name (str): The basis indicating the addressed atoms after
                 the pulse sequence ('ground-rydberg', 'digital' or 'all').
             meas_basis (Optional[str]): The basis in which a sampling
                 measurement is desired.
+            sim_times (array): Array of times (µs) when simulation results are
+                returned.
         """
-        self._states = run_output
         self._dim = dim
         self._size = size
         if basis_name not in {'ground-rydberg', 'digital', 'all'}:
             raise ValueError(
                 "`basis_name` must be 'ground-rydberg', 'digital' or 'all'."
                 )
         self._basis_name = basis_name
+        self.sim_times = sim_times
+        self._results: Union[list[Counter], list[qutip.Qobj]]
+        self._weights: list[float]
+
+    @abstractmethod
+    def get_state(self, t: int) -> qutip.Qobj:
+        """Returns the state of the system at time t."""
+        pass
+
+    @abstractmethod
+    def get_final_state(self) -> qutip.Qobj:
+        """Returns the final state of the system."""
+        pass
+
+    @abstractmethod
+    def sample_final_state(self, N_samples: int = 1000) -> Counter:
+        """Returns the result of multiple measurements of the final state."""
+        pass
```

Comment:
Same for this one, since you have `sample_state`?

Reply:
If you specify `_calc_weights` for both classes, then the definition of `sample_state` can be in the parent class as:
```
        self. _calc_weights()
        return super().sample_state(t, N_samples)
```
something like that. You see what I mean? You might not even have to define `sample_state` outside of the parent class.

Reply:
Yep, I thought about doing this, but there is the problem that both child classes don't have `meas_basis` as an argument, and so both classes have a different method signature... Is there a way around this ?

Reply:
I mean, you need to specify a `meas_basis` for `CleanResults._calc_weight`, but not for `NoisyResults._calc_weight`.

Reply:
What about having `sample_state` and `_calc_weight` accept `*args` and `**kwargs`?